### DISCUSSION TITLE
Letsencrypt Pre-Flight 404

### DISCUSSION THREAD

**Discussion by breaker-of-stone** [requarks/wiki#3209] [Help / Questions] [answered]
The Letsencrypt pre-flight curl fetch gets a 404, and I notice that the quoted curl omits the port (:3000).

1.  Wiki.js working normally on default port.  No reverse proxy, wiki.js connecting directly to the port.
2. Modified config.yml to require Letsencrypt as follows:

```
# ---------------------------------------------------------------------
# Port the server should listen to
# ---------------------------------------------------------------------

port: 3000

# ---------------------------------------------------------------------
# SSL/TLS Settings
# ---------------------------------------------------------------------
# Consider using a reverse proxy (e.g. nginx) if you require more
# advanced options than those provided below.

ssl:
  enabled: true
  port: 3443
  provider: letsencrypt

  # ++++++ For letsencrypt only ++++++
  domain: wiki.noisebin.org
  subscriberEmail: admin@noisebin.org

#   # Provider to use, possible values: custom, letsencrypt
#   provider: custom
# 
#   # ++++++ For custom only ++++++
#   # Certificate format, either 'pem' or 'pfx':
#   format: pem
#   # Using PEM format:
#   key: path/to/key.pem
#   cert: path/to/cert.pem
#   # Using PFX format:
#   pfx: path/to/cert.pfx
#   # Passphrase when using encrypted PEM / PFX keys (default: null):
#   passphrase: null
#   # Diffie Hellman parameters, with key length being greater or equal
#   # to 1024 bits (default: null):
#   dhparam: null
```

### Result:

```
2021-03-13T07:28:41.479Z [MASTER] warn: (LETSENCRYPT) Error: Error: Failed HTTP-01 Pre-Flight / Dry Run.
curl 'http://wiki.noisebin.org/.well-known/acme-challenge/test-2d1300ab355ba41aac2cb2c1b3790db6-0'
Expected: 'test-2d1300ab355ba41aac2cb2c1b3790db6-0.EvusP9a9zkFE3M3-_bM2eBpdLWV8SxdBPVEev0M76ok'
Got: '<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.14.0 (Ubuntu)</center>
</body>
</html>
'

```
Should there be a :3000 in that curl request?

How?

**Comment by NGPixel** [comment, marked as answer]
You can't use port 3000. Either use 443 or put a reverse proxy like nginx in front of Wiki.js if you need to a more advanced configuration with a non-standard port.

**Comment by breaker-of-stone** [comment]
I accept but don't understand the principles behind the response. 
The documentation doesn't explain these dependencies, although I can appreciate they follow convention.  
The curl request target (implied :80) is also not explained.
The build process doesn't give a useful message.

My solution:
 * nginx is already installed and in use for other components of the site
 * run certbot in offline mode
 * alter nginx/sites-enabled/this-site so it:
   * uses the generated certificate and key
   * connects to the wiki.js instance via localhost:3000 (reverse proxy mode)
   * add a block that redirects http to https
 
All AOK.

**Reply by NGPixel** [reply]
The built-in Let's Encrypt feature is meant for basic setups (HTTP and HTTPS on standard ports). It was never meant to address all the possible scenarios (non-standard ports or DNS verification method).

**Comment by mattpr** [comment]
> I accept but don't understand the principles behind the response.

It is a certbot/letsencrypt error and doesn't really have anything to do with wiki.js.  To do the domain validation (letsencrypt needs to verify you control the domain before signing a certificate for that domain for you) letsencrypt tries to find a well-known file on your domain under port 80 (http).  It will follow redirects so if you redirect from 80 to 443 (ssl) that will work as long as you already have an already-working https setup (certificates in place).

I believe letsencrypt won't let you provide random ports because then if you had a server get hacked, the hacker could run a little webserver on non-priv ports (e.g. 3000) and get a signed certificate for your domain and then do more damage (man in the middle attacks, etc).

But as I said this is out of scope of wiki.js so you would best look into certbot/letsencrypt and their rules/logic around how domain validation works.

Bottom line is that you need to answer their validation request on normal http port in order to get the cert generated.

Looks like NGPixel beat me to answer.  :)
There is more detail about letsencrypt, what can go wrong and my setup here if you are looking for more info.
https://github.com/Requarks/wiki/discussions/3040

**Reply by breaker-of-stone** [reply]
Belated thanks for a thorough explanation!
